README.md
# QuickButton
[![pub - v0.0.1](https://img.shields.io/badge/pub-v0.0.1-blue)](https://) [![unit test - passing](https://img.shields.io/badge/unit_test-passing-2ea44f?logo=github)](https://) [![license - MIT](https://img.shields.io/badge/license-MIT-blue)](https://opensource.org/license/mit)

A customizable button widget for Flutter applications that allows you to easily create buttons with various attributes and behaviors.

![Demo](https://firebasestorage.googleapis.com/v0/b/mocon-3075b.appspot.com/o/14.08.2023_01.18.45_REC.gif?alt=media&token=b80d9653-a66b-4921-8ab7-3b5c7ef5aa50)

## Resources

- [Pub package](https://pub.dev/packages/quick_button).
 - [Github](https://github.com/pozadkey/quick_button).

## Installation

To use this package, add `quick_button` as a dependency in your `pubspec.yaml` file:

```yaml
dependencies:
  flutter:
    sdk: flutter
  quick_button: ^1.0.0 # Use the latest version
```

## Usage

Import the package in your Dart code:

```dart
import 'package:quick_button/quick_button.dart';
```

You can then use the `QuickButton` widget to create buttons with different variations:

```dart
QuickButton(
  labelText: 'Submit',
  onPressed: () {
    // Action to perform when the button is pressed
  },
)

QuickButton.icon(
  icon: Icons.star,
  onPressed: () {
    // Action to perform when the button is pressed
  },
)

QuickButton.prefixIcon(
  labelText: 'Start',
  prefixIcon: Icons.play_arrow,
  onPressed: () {
    // Action to perform when the button is pressed
  },
)

QuickButton.suffixIcon(
  labelText: 'End',
  suffixIcon: Icons.stop,
  onPressed: () {
    // Action to perform when the button is pressed
  },
)
```

## Attributes

- `labelText`: The label text displayed on the button.
- `labelColor`: The color of the label text.
- `labelSize`: The font size of the label text.
- `labelWeight`: The font weight of the label text.
- `labelSpacing`: The spacing between characters in the label text.
- `icon`: The icon placed within the button.
- `prefixIcon`: The icon placed before the label text.
- `suffixIcon`: The icon placed after the label text.
- `iconSize`: The size of all icons. Default value is `15.0`.
- `buttonWidth`: The width of the button. Default value is `150.0`.
- `buttonHeight`: The height of the button.
- `borderRadius`: The degree of the roundness for the corners of the button.
- `borderColor`: The color of the border surrounding the button.
- `borderSize`: The width of the border surrounding the button. Default value is `1.0`.
- `backgroundColor`: The background color of the button.
- `hoverIn`: The color of the label text when the mouse moves within the widget.
- `backgroundHoverIn`: The background color of button when the mouse moves within the widget.
- `hoverOut`: The color of the label text when the mouse moves outside the widget.
- `backgroundHoverOut`: The background color of button when the mouse moves outside the widget.
- `onPressed`: Callback function to execute when the button is pressed.



## Callback

All variations of the `QuickButton` widget require an `onPressed` callback function that gets triggered when the button is pressed.

## License

This package is released under the [MIT License](https://opensource.org/license/mit).

## Contributions

Contributions are welcome! If you find any issues or want to enhance this package, feel free to submit a pull request [here](https://github.com/pozadkey/quick_button).
 



CHANGELOG.md


## [1.0.0] - 2023-08-17
- Added `QuickButton` widget with various attributes and customization options.
- Added Four variations of the `QuickButton` widget: default, icon, prefixIcon, and suffixIcon.
- Fixed documentation errors.

## [0.1.2] - 2023-08-15

* Restructed `CHANGELOG.md`.

## [0.1.1] - 2023-08-15
* Fixed Documentation.

## [0.1.0] - 2023-08-15

* Depreciated **`buttonSize`** property.
* Added **`icon`**, **`buttonWidth`** & **`buttonHeight`** properties.
* Added Documentation.
* Updated ReadME.md.

## [0.0.1] - 2023-08-14 

* First release.






test/quick_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:quick_button/quick_button.dart';

void main() {
  testWidgets('Click button', (WidgetTester tester) async {
    var pressed = false;
    final enterButton = find.byKey(const ValueKey('onTapButton'));

    await tester.pumpWidget(MaterialApp(
        home: Scaffold(
      body: QuickButton(
          labelText: 'Hello',
          onPressed: () {
            pressed = true;
          }),
    )));
    await tester.tap(enterButton);
    await tester.pump();
    expect(pressed, true);
  });
}


test/test_material_app_widget.dart
// ignore_for_file: must_be_immutable

import 'package:flutter/material.dart';

class TestMaterialAppWidget extends StatelessWidget {
  Widget? home;

  TestMaterialAppWidget({super.key, this.home});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Widget Test',
      home: home,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:quick_button/quick_button.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Quick Button'),
          backgroundColor: Colors.purple,
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              QuickButton(
                labelText: 'Submit',
                labelColor: Colors.purple,
                labelSize: 13.0,
                labelWeight: FontWeight.bold,
                labelSpacing: 0.3,
                borderColor: Colors.purple,
                borderSize: 1.5,
                backgroundColor: Colors.transparent,
                hoverIn: Colors.white,
                hoverOut: Colors.purple,
                backgroundHoverIn: Colors.purple,
                backgroundHoverOut: Colors.transparent,
                onPressed: () {
                  // Action to perform when the button is pressed
                },
              ),
              const SizedBox(
                height: 15,
              ),
              QuickButton.icon(
                icon: Icons.star,
                labelColor: Colors.purple,
                buttonHeight: 30.0,
                buttonWidth: 30.0,
                borderColor: Colors.purple,
                borderSize: 1.5,
                backgroundColor: Colors.transparent,
                hoverIn: Colors.white,
                hoverOut: Colors.purple,
                backgroundHoverIn: Colors.purple,
                backgroundHoverOut: Colors.transparent,
                onPressed: () {
                  // Action to perform when the button is pressed
                },
              ),
              const SizedBox(
                height: 15,
              ),
              QuickButton.prefixIcon(
                labelText: 'Quick Button',
                labelColor: Colors.white,
                prefixIcon: Icons.play_arrow,
                borderRadius: 25.0,
                borderColor: Colors.transparent,
                borderSize: 1.5,
                labelWeight: FontWeight.w700,
                backgroundColor: Colors.purple,
                hoverIn: Colors.purple[800],
                hoverOut: Colors.white,
                backgroundHoverIn: const Color.fromARGB(162, 155, 39, 176),
                backgroundHoverOut: Colors.purple,
                onPressed: () {
                  // Action to perform when the button is pressed
                },
              ),
              const SizedBox(
                height: 15,
              ),
              QuickButton.suffixIcon(
                labelText: 'End',
                labelColor: Colors.purple,
                suffixIcon: Icons.stop,
                borderRadius: 5.0,
                borderColor: Colors.purple,
                borderSize: 1.5,
                backgroundColor: Colors.transparent,
                hoverIn: Colors.white,
                hoverOut: Colors.purple,
                backgroundHoverIn: Colors.purple,
                backgroundHoverOut: Colors.transparent,
                onPressed: () {
                  // Action to perform when the button is pressed
                },
              ),
              const SizedBox(
                height: 15,
              ),
              QuickButton.icon(
                icon: Icons.book,
                labelColor: Colors.white,
                buttonHeight: 30.0,
                buttonWidth: 30.0,
                borderRadius: 25.0,
                borderColor: Colors.purple,
                borderSize: 1.5,
                backgroundColor: Colors.purple,
                hoverIn: Colors.purple,
                hoverOut: Colors.white,
                backgroundHoverIn: Colors.white,
                backgroundHoverOut: Colors.purple,
                onPressed: () {
                  // Action to perform when the button is pressed
                },
              ),
              const SizedBox(
                height: 15,
              ),
              QuickButton(
                labelText: 'Link',
                labelColor: Colors.purple,
                hoverIn: Colors.black,
                hoverOut: Colors.purple,
                backgroundColor: Colors.white,
                onPressed: () {
                  // Action to perform when the button is pressed
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/quick_button.dart
/// A library that provides the [QuickButton] widget for creating customizable buttons.

library quick_button;

// ignore_for_file: must_be_immutable

import 'package:flutter/material.dart';

/// A customizable button widget for Flutter applications.
class QuickButton extends StatefulWidget {
  /// The label text displayed on the button.
  final String? labelText;

  /// The color of the label text. Default is Colors.white.
  Color? labelColor;

  /// The font size of the label text. Default size is 13.
  final double? labelSize;

  /// The font weight of the label text. Default is FontWeight.w600
  final FontWeight? labelWeight;

  /// The spacing between characters in the label text.
  final double? labelSpacing;

  /// The icon placed within the button.
  final IconData? icon;

  /// The icon placed before the label text.
  final IconData? prefixIcon;

  /// The icon placed after the label text.
  final IconData? suffixIcon;

  /// The size of all icons. Default value is 15.0.
  final double? iconSize;

  /// The width of the button. Default value is 150.0.
  final double? buttonWidth;

  /// The height of the button.
  final double? buttonHeight;

  /// The degree of the roundness for the corners of the button.
  final double? borderRadius;

  /// The color of the border surrounding the button.
  final Color? borderColor;

  /// The width of the border surrounding the button. Default value is 1.0.
  final double? borderSize;

  /// The background color of the button.
  Color? backgroundColor;

  /// The color of the label text when the mouse moves within the widget.
  final Color? hoverIn;

  /// The background color of button when the mouse moves within the widget.
  final Color? backgroundHoverIn;

  /// The color of the label text when the mouse moves outside the widget.
  final Color? hoverOut;

  /// The background color of button when the mouse moves outside the widget.
  final Color? backgroundHoverOut;

  /// Callback function to execute when the button is pressed.
  final Function onPressed;

  /// Creates a new [QuickButton] instance.
  ///
  /// The [onPressed] callback must be provided to handle the button tap event.
  QuickButton({
    Key? key,
    required this.labelText,
    this.labelColor,
    this.labelSize,
    this.labelWeight,
    this.labelSpacing,
    this.buttonWidth,
    this.buttonHeight,
    this.borderRadius,
    this.borderColor,
    this.borderSize,
    this.backgroundColor,
    this.hoverIn,
    this.backgroundHoverIn,
    this.hoverOut,
    this.backgroundHoverOut,
    required this.onPressed,
  })  : icon = null,
        prefixIcon = null,
        suffixIcon = null,
        iconSize = null,
        super(key: key);

  /// Creates a new [QuickButton] instance with only an icon.
  ///
  /// The [icon] parameter specifies the icon placed within the button.
  /// The [onPressed] parameter is required and specifies the callback function to be invoked when the button is pressed.
  QuickButton.icon({
    Key? key,
    required this.icon,
    this.labelColor,
    this.iconSize,
    this.buttonWidth,
    this.buttonHeight,
    this.borderRadius,
    this.borderColor,
    this.borderSize,
    this.backgroundColor,
    this.hoverIn,
    this.backgroundHoverIn,
    this.hoverOut,
    this.backgroundHoverOut,
    required this.onPressed,
  })  : labelText = null,
        labelSize = null,
        labelSpacing = null,
        labelWeight = null,
        prefixIcon = null,
        suffixIcon = null,
        super(key: key);

  /// Creates a new [QuickButton] instance with only a prefix icon.
  ///
  /// The [labelText] parameter specifies the text label to display on the button.
  /// The [prefixIcon] parameter specifies the icon placed before the label text.
  /// The [onPressed] parameter is required and specifies the callback function to be invoked when the button is pressed.
  QuickButton.prefixIcon({
    Key? key,
    required this.labelText,
    this.labelColor,
    this.labelSize,
    this.labelWeight,
    this.labelSpacing,
    required this.prefixIcon,
    this.iconSize,
    this.buttonWidth,
    this.buttonHeight,
    this.borderRadius,
    this.borderColor,
    this.borderSize,
    this.backgroundColor,
    this.hoverIn,
    this.backgroundHoverIn,
    this.hoverOut,
    this.backgroundHoverOut,
    required this.onPressed,
  })  : icon = null,
        suffixIcon = null,
        super(key: key);

  /// Creates a new [QuickButton] instance with only a suffix icon.
  ///
  /// The [labelText] parameter specifies the text label to display on the button.
  /// The [suffixIcon] parameter specifies the icon placed after the label text.
  /// The [onPressed] parameter is required and specifies the callback function to be invoked when the button is pressed.
  QuickButton.suffixIcon({
    Key? key,
    required this.labelText,
    this.labelColor,
    this.labelSize,
    this.labelWeight,
    this.labelSpacing,
    required this.suffixIcon,
    this.iconSize,
    this.buttonWidth,
    this.buttonHeight,
    this.borderRadius,
    this.borderColor,
    this.borderSize,
    this.backgroundColor,
    this.hoverIn,
    this.backgroundHoverIn,
    this.hoverOut,
    this.backgroundHoverOut,
    required this.onPressed,
  })  : icon = null,
        prefixIcon = null,
        super(key: key);

  @override
  State<QuickButton> createState() => _QuickButtonState();
}

class _QuickButtonState extends State<QuickButton> {
  @override
  Widget build(BuildContext context) {
    // Obtains the width of the current screen's viewport.
    double width = MediaQuery.of(context).size.width;

    return MouseRegion(
      /// Detects mouse movement within the widget.
      onEnter: (_) {
        _updateButtonColors(widget.hoverIn, widget.backgroundHoverIn);
      },

      /// Detects mouse movement outside the widget.
      onExit: (_) {
        _updateButtonColors(widget.hoverOut, widget.backgroundHoverOut);
      },

      child: InkWell(
        key: const Key('onTapButton'),
        child: Container(
          key: const Key('buttonDecoration'),
          width: widget.buttonWidth ?? 150.0,
          height: widget.buttonHeight ?? 35.0,
          padding: width >= 800
              ? widget.labelText == null
                  ? const EdgeInsets.all(4)
                  : const EdgeInsets.all(6)
              : const EdgeInsets.all(5),
          alignment: Alignment.center,
          decoration: BoxDecoration(
              color: widget.backgroundColor ?? Colors.transparent,
              borderRadius: BorderRadius.circular(
                widget.borderRadius ?? 0.0,
              ),
              border: Border.all(
                  color: widget.borderColor ?? Colors.transparent,
                  width: widget.borderSize ?? 1.0)),
          child: FittedBox(
            child: Row(
              children: _buildButtonChildren(),
            ),
          ),
        ),
      ),
    );
  }

  /// Updates the button's label color and background color.
  ///
  /// The [labelColor] parameter specifies the new label color.
  /// The [backgroundColor] parameter specifies the new background color.
  void _updateButtonColors(Color? labelColor, Color? backgroundColor) {
    setState(() {
      widget.labelColor = labelColor;
      widget.backgroundColor = backgroundColor;
    });
  }

  /// Builds and returns the list of child widgets for the button.
  List<Widget> _buildButtonChildren() {
    final buttonFont = TextStyle(
        fontSize: widget.labelSize,
        color: widget.labelColor ?? Colors.black,
        letterSpacing: widget.labelSpacing ?? 0.3,
        fontWeight: widget.labelWeight ?? FontWeight.w600);

    return [
      if (widget.suffixIcon == null && widget.prefixIcon != null) ...{
        const SizedBox(
          width: 18,
        ),
        Icon(
          widget.prefixIcon,
          color: widget.labelColor,
          size: widget.iconSize ?? 15,
        ),
        const SizedBox(
          width: 7,
        ),
        widget.labelText == null
            ? Icon(
                widget.icon,
                size: widget.iconSize,
                color: widget.labelColor,
              )
            : Text(
                key: const Key('labelText'),
                widget.labelText.toString(),
                style: buttonFont,
              ),
        Icon(
          widget.suffixIcon,
          color: widget.labelColor,
          size: widget.iconSize ?? 15,
        ),
      } else if (widget.prefixIcon == null && widget.suffixIcon != null) ...{
        Icon(
          widget.prefixIcon,
          color: widget.labelColor,
          size: widget.iconSize ?? 15,
        ),
        widget.labelText == null
            ? Icon(
                widget.icon,
                size: widget.iconSize,
                color: widget.labelColor,
              )
            : Text(
                key: const Key('labelText'),
                widget.labelText.toString(),
                style: buttonFont,
              ),
        const SizedBox(
          width: 7,
        ),
        Icon(
          widget.suffixIcon,
          color: widget.labelColor,
          size: widget.iconSize ?? 15,
        ),
        const SizedBox(
          width: 18,
        ),
      } else if (widget.prefixIcon == null || widget.suffixIcon == null) ...{
        widget.labelText == null
            ? Icon(
                widget.icon,
                size: widget.iconSize,
                color: widget.labelColor,
              )
            : Text(
                key: const Key('labelText'),
                widget.labelText.toString(),
                style: buttonFont,
              ),
      } else if (widget.prefixIcon == null ||
          widget.suffixIcon == null ||
          widget.labelText == null) ...{
        widget.labelText == null
            ? Icon(
                widget.icon,
                size: widget.iconSize,
                color: widget.labelColor,
              )
            : Text(
                key: const Key('labelText'),
                widget.labelText.toString(),
                style: buttonFont,
              ),
      }
    ];
  }
}
